Chapter11

danieljharris opened the pull request.

danieljharris committed:
Completed Exercise 11-1

danieljharris committed:
Finished Chapter11

danieljharris committed:
Added notes to das_tester

danieljharris committed:
Removed tabs

danieljharris committed:
Removed tabs

danieljharris committed:
Removed Tabs

danieljharris committed:
Removed tabs

danieljharris committed:
Removed tabs

velimir commented on Chapter11/balance_node.erl:
```diff
@@ -0,0 +1,68 @@
+    
+    {lookup, Address, Pid}    ->
+      event_manager:send_event(node_logger, {lookup, {Address}, NextNode}),
+      clear_mailbox(),
```
why do you need to clear mailbox here?

velimir commented on Chapter11/das.erl:
```diff
@@ -0,0 +1,47 @@
+remove(Address) -> {balancer, ?BALANCE_NODE} ! {remove, Address}.
+
+lookup(Address)	->
+  clear_mailbox(),
```
those clear mailboxes confuse me

velimir commented on Chapter11/das_node.erl:
```diff
@@ -0,0 +1,53 @@
+
+    {link_node, Node} ->
+      % net_kernel:connect(Node),
+      monitor_node(Node, true),
```
It doesn't look like monitor is used.  I don't see that process can receive `nodedown` messages and handle them.

velimir commented on Chapter11/das_node.erl:
```diff
@@ -0,0 +1,53 @@
+      address_loop();
+
+    {lookup, Address, Pid} ->
+      clear_mailbox(),
```
Clear mailbox again. This doesn't look right.

velimir reviewed (changes requested):
Can you please explain why you need to clear mailboxes? If this is necessary, then it's a red flag that there's a bug in code.
Also your solution has a bottleneck - all read and write operations go through the single DB process, whilst book says that DB has to be replicated (so all nodes have data). Reads go to single node (chosen by balancer randomly), Writes go to all nodes.

danieljharris committed:
Fixed clear_mailbox()

Fixed the problem that caused me to need clear_mailbox() and removed the now un-needed clear_mailbox() function

danieljharris committed:
2 db store nodes!

Updated the code to allow for db store nodes to be added. I've implemented 2 db nodes but the code allows for more to be added easily

danieljharris committed:
Removed node crash recovery

I removed the not working linking and monitoring as it was not required by the exercises

velimir commented on Chapter11/das_node.erl:
```diff
@@ -0,0 +1,39 @@
+-module(das_node).
+-export([create/1, address_loop/1]).
```
now I'm confused why das node is needed? it seems like you can just replace its functionality by adding some functionality to balancer node, so that write/read/delete messages do not go through several processes, but rather directly from balancer to node.

danieljharris replied:
Ahh yeah I see what you mean, I've deleted the das_node nodes and shifted there functionality to balancer. I also changed how the lookup works, so instead of the balance node waiting for a response from the a db_node, the db_node sends the lookup response straight to the das:lookup function :)

danieljharris replied:
The lookup change should reduce bottlenecks in balancer

danieljharris committed:
Removed das_node and made improvements
